### PR TITLE
DR-792 Use templating library to generate BigQuery SQL blocks

### PR DESCRIPTION
**Pull request by danxmoran**
There's still a fair amount of `+`-ing, but it's nearly all in `static` members so I'm not too worried about performance implications. Hopefully the team agrees this is an improvement!

### CODE REVIEW

File: src/main/java/bio/terra/common/PdaoConstant.java [hunk 1]
```diff
@@ -14,4 +11,7 @@ private PdaoConstant() {
     public static final String STAGING_TABLE_ROW_ID_COLUMN = PDAO_PREFIX + STAGING_TABLE_PREFIX + "row_id";
     public static final String TARGET_TABLE_ROW_ID_COLUMN = PDAO_PREFIX + TARGET_TABLE_PREFIX + "row_id";
 
+    public static String prefixName(String name) {
```
Comment (ddietterich) [inline src/main/java/bio/terra/common/PdaoConstant.java]: Not very constant. Maybe rename this module to PdaoNames.

File: src/main/java/bio/terra/service/tabulardata/google/BigQuerySqlGenerator.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
+package bio.terra.service.tabulardata.google;
```
Comment (ddietterich) [inline src/main/java/bio/terra/service/tabulardata/google/BigQuerySqlGenerator.java]: I like the templating a lot! I'm not seeing a lot of value to factoring this out into a separate module, though.

My idea was to rearrange the Pdao by operations (dataset, snapshot, ingest) with a shared "utils" module. FWIW

**Comment by danxmoran** [comment]
i/o timeout trying to connect to k8s in Travis

**Review by ddietterich — COMMENTED** [review]
Love the templating.
Not keen on the SQL generation module.

File: src/test/java/bio/terra/service/tabulardata/google/BigQueryPdaoTest.java [hunk 3]
```diff
@@ -41,7 +41,7 @@
 import org.springframework.test.context.ActiveProfiles;
 import org.springframework.test.context.junit4.SpringRunner;
 
-import static bio.terra.common.PdaoConstant.PDAO_ROW_ID_COLUMN;
+import static bio.terra.common.PdaoConstant.*;
```
Comment (thathert) [inline src/test/java/bio/terra/service/tabulardata/google/BigQueryPdaoTest.java]: We are anti-asterisk here *️⃣ 🚫

**Review by fboulnois — COMMENTED** [review]
The string templating looks nice. Was the idea behind separating the SQL into a separate module to make it easier to test?

**Comment by danxmoran** [comment]
> Was the idea behind separating the SQL into a separate module to make it easier to test?

That + trying to shrink the size / responsibility of the PDAO class. I am convinced by the counter-proposal of splitting the PDAO by operations, but I don't think I'll bite that off here. When I go through to port the other queries to use the templating, I'll move these two back into the PDAO.

File: src/main/java/bio/terra/common/PdaoConstant.java (hunk 1 above)
Comment (danxmoran) [reply src/main/java/bio/terra/common/PdaoConstant.java]: Was only needed when splitting out the new file, will move it back to the PDAO

**Comment by danxmoran** [comment]
@ddietterich @thathert @fboulnois this is ready for another look. The Travis <-> k8s links is unhappy again, but `BigQueryPdaoTest` and `EncodeFileTest` both pass for me locally.

**Review by fboulnois — COMMENTED** [review]
lgtm

**Comment by pshapiro4broad** [comment]
I haven’t looked at this closely but have you looked at bigquery’s support for parameterized queries?

Such as: https://cloud.google.com/bigquery/docs/parameterized-queries#using_arrays_in_parameterized_queries

**Comment by danxmoran** [comment]
> have you looked at bigquery’s support for parameterized queries?

I hadn't looked at it before, but I don't think it's an equivalent replacement for the string-templating used here. The parameters can't be used for substituting identifiers or column/table names, which is where we get a lot of value from the templating.

I could see a follow-up ticket to use the parameterized queries in the places where it's possible, to guard against SQL injection.

**Review by RoriCremer — APPROVED** [review]

**Comment by pshapiro4broad** [comment]
> I hadn't looked at it before, but I don't think it's an equivalent replacement for the string-templating used here. The parameters can't be used for substituting identifiers or column/table names, which is where we get a lot of value from the templating.

In their examples that they use `String.format()` to perform table name etc substitution and then use the parameter APIs to substitute query values. So a mixed approach may be best.

**Pull request merged**